pkg/volume/csi/csi_attacher_test.go

```diff
@@ -1023,23 +1024,40 @@ func TestAttacherGetDeviceMountPath(t *testing.T) {
 	csiAttacher := getCsiAttacherFromVolumeAttacher(attacher, testWatchTimeout)
 
 	pluginDir := csiAttacher.plugin.host.GetPluginDir(plug.GetPluginName())
-
 	testCases := []struct {
-		testName          string
-		pvName            string
-		expectedMountPath string
-		shouldFail        bool
+		testName           string
+		pvName             string
+		volumeId           string
+		skipPVCSISource    bool // The test clears PV.Spec.CSI
+		shouldFail         bool
+		addVolSource       bool // The test adds a Volume.VolumeSource.CSI.
+		removeVolumeHandle bool // The test force removes CSI volume handle.
 	}{
 		{
-			testName:          "normal test",
-			pvName:            "test-pv1",
-			expectedMountPath: pluginDir + "/pv/test-pv1/globalmount",
+			testName: "success test",
+			pvName:   "test-pv1",
+			volumeId: "test-vol1",
+		},
+		{
+			testName:        "fail test, failed to create device mount path due to missing CSI source",
+			pvName:          "test-pv1",
+			volumeId:        "test-vol1",
+			skipPVCSISource: true,
+			shouldFail:      true,
+		},
+		{
+			testName:     "fail test, failed to create device mount path, CSIVolumeSource found",
```

Comment:
why CSIVolumeSource found will fail the test?

Reply:
This is because csi_attacher.go uses getPVSourceFromSpec function fails if CSIVolumeSource is detected. [code](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/csi/csi_util.go#L171)